src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,226 @@
+                pagesRead, pagesEmpty, recentPagesRead, recentPagesEmpty, batchLastPosition, lastCommittedPosition, batchCaughtUp, currentBatches, maxBatches)
+
+        member this.UpdateBatch (batch: InternalBatch) =
+            batchFirstPosition <- batch.firstPosition
+            batchLastPosition <- batch.lastPosition
+            batchCaughtUp <- batch.isEnd
+
+            pagesRead <- pagesRead + 1
+            recentPagesRead <- recentPagesRead + 1
+
+        member this.UpdateEmptyPage () =
+            pagesEmpty <- pagesEmpty + 1
+            recentPagesEmpty <- recentPagesEmpty + 1
+
+        member this.UpdateCommitedPosition(pos) =
+            lastCommittedPosition <- pos
+
+        member this.UpdateCurMax(cur, max) =
+            currentBatches <- cur
+            maxBatches <- max
+
+        member this.Start =
+            async {
+                let! ct = Async.CancellationToken
+                while not ct.IsCancellationRequested do
+                    report ()
+                    recentPagesRead <- 0
+                    recentPagesEmpty <- 0
+                    do! Async.Sleep statsInterval
+            }
+
+    [<RequireQualifiedAccess>]
+    [<NoComparison>]
+    type Work =
+        | TakeInitial
+        | Page of ReadAllPage
+        | TakeNext of ReadAllPage
+
+    type SubmitBatchHandler =
+        // ingester submit method: epoch * checkpoint * items -> write result
+        int64 * Async<unit> * seq<Propulsion.Streams.StreamEvent<byte[]>> -> Async<int*int>
+
+type StreamReader
+    (
+        logger : ILogger,
+        store: IStreamStore,
+        ledger: ICheckpointer,
+        submitBatch: SubmitBatchHandler,
+        streamId,
+        consumerGroup,
+        ?maxBatchSize: int,
+        ?sleepInterval: TimeSpan,
+        ?statsInterval: TimeSpan,
+        ?stopAfterInterval: TimeSpan
+    ) =
+
+    let maxBatchSize = defaultArg maxBatchSize 100
+    let sleepInterval = defaultArg sleepInterval (TimeSpan.FromSeconds(5.))
+    let stopAfterInterval = defaultArg stopAfterInterval TimeSpan.MaxValue
+
+    let stats = Stats(logger, ?statsInterval = statsInterval)
+
+    let commit batch =
+        async {
+            try
+                do! ledger.CommitPosition { Stream = streamId; ConsumerGroup = consumerGroup; Position = Nullable(batch.lastPosition) }
+                stats.UpdateCommitedPosition(batch.lastPosition)
+                logger.Debug("Committed position {position}", batch.lastPosition)
+            with
+            | exc ->
+                logger.Error(exc, "Exception while commiting position {position}", batch.lastPosition)
+                return! Async.Raise exc
+        }
+
+    let processPage (page: ReadAllPage) =
+        async {
+            if page.Messages.Length > 0 then
+
+                let events =
+                    page.Messages
+                    |> Seq.map StreamMessage.intoStreamEvent
+                    |> Seq.sortBy (fun x -> x.event.Index)
```

Comment:
why is this necessary?

Reply:
Just to make sure the logs come out right, I don't assume page.Messages are ordered, even if they probably are.

Reply:
I believe based on what I know of SSS (and how it works in ES) that they are guaranteed to be in order of Position
However I believe that (see other comment re conflating Index and Position) this would end up ordering them by ordinal in the stream, which would not make sense

Reply:
Yeah, that needs to change or go away when the index/position change happens.